File: src/main/kotlin/top_75/_1493_longest_subarray_of_1s_after_deleting_one_element.java
package top_75;

public class _1493_longest_subarray_of_1s_after_deleting_one_element {
    public int longestSubarray(int[] nums) {
        int l = 0;
        int zeroCnt = 0;
        int ans = 0;
        for (int r = 0; r < nums.length; r++) {
            if (nums[r] == 0) {
                zeroCnt++;
            }
            if (zeroCnt > 1) {
                if (nums[l] == 0) zeroCnt--;
                l++;
            }
            if (zeroCnt <= 1) {
                ans = Math.max(ans, r - l);
            }
        }
        return ans;
    }

    public static void main(String[] args) {
        var s = new _1493_longest_subarray_of_1s_after_deleting_one_element();
        System.out.println(s.longestSubarray(new int[]{1, 1, 0, 1}));
        System.out.println(s.longestSubarray(new int[]{0, 1, 1, 1, 0, 1, 1, 0, 1}));
        System.out.println(s.longestSubarray(new int[]{1, 1, 1}));
    }
}

File: src/main/kotlin/topic/slidingwindow/medium/_1493_longest_subarray_of_1_s_after_deleting_one_element.kt
package topic.slidingwindow.medium

import kotlin.math.max

class _1493_longest_subarray_of_1_s_after_deleting_one_element {
    fun longestSubarray(nums: IntArray): Int {
        var l = 0
        var r = 0
        var max = 0
        var init = false
        for (i in nums.indices) {
            if (!init) {
                if (nums[i] == 1) {
                    l++
                } else {
                    init = true
                }
            } else {
                if (nums[i] == 1) {
                    r++
                } else {
                    max = max(max, l + r)
                    l = r
                    r = 0
                }
            }
        }
        if (!init) return nums.size - 1
        max = max(max, l + r)
        return max
    }
}

fun main() {
    with(_1493_longest_subarray_of_1_s_after_deleting_one_element()) {
        println(longestSubarray(intArrayOf(1,1,0,1)))
        println(longestSubarray(intArrayOf(0,1,1,1,0,1,1,0,1)))
        println(longestSubarray(intArrayOf(1,1,1)))
    }
}
